syncer/syncer.go

```diff
@@ -37,58 +36,38 @@ const (
 	eventWarning = "Warning"
 )
 
-// Given an Interface, returns a controllerutil.MutateFn which also sets the
-// owner reference if the subject has one
-func syncFn(syncer Interface, scheme *runtime.Scheme) controllerutil.MutateFn {
-	owner := syncer.GetOwner()
-	return func(existing runtime.Object) error {
-		err := syncer.SyncFn(existing)
-		if err != nil {
-			return err
-		}
-		if owner != nil {
-			existingMeta, ok := existing.(metav1.Object)
-			if !ok {
-				return fmt.Errorf("%T is not a metav1.Object", existing)
-			}
-			ownerMeta, ok := owner.(metav1.Object)
-			if !ok {
-				return fmt.Errorf("%T is not a metav1.Object", owner)
-			}
-			err := controllerutil.SetControllerReference(ownerMeta, existingMeta, scheme)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
+func getKey(obj runtime.Object) (types.NamespacedName, error) {
+	key := types.NamespacedName{}
+	objMeta, ok := obj.(metav1.Object)
+	if !ok {
+		return key, fmt.Errorf("%T is not a metav1.Object", obj)
 	}
+
+	key.Name = objMeta.GetName()
+	key.Namespace = objMeta.GetNamespace()
+	return key, nil
 }
 
-// Sync mutates the subject of the syncer interface using controller-runtime
-// CreateOrUpdate method. It takes care of setting owner references and
-// recording kubernetes events where appropriate
-func Sync(ctx context.Context, syncer Interface, c client.Client, scheme *runtime.Scheme, recorder record.EventRecorder) error {
-	obj := syncer.GetObject()
-	objMeta, ok := obj.(metav1.Object)
-	if !ok {
-		return fmt.Errorf("%T is not a metav1.Object", obj)
+func basicEventReason(objKindName string, err error) string {
+	if err != nil {
+		return fmt.Sprintf("%sSyncFailed", strcase.ToCamel(objKindName))
 	}
-	key := types.NamespacedName{Name: objMeta.GetName(), Namespace: objMeta.GetNamespace()}
-	op, err := controllerutil.CreateOrUpdate(ctx, c, obj, syncFn(syncer, scheme))
+	return fmt.Sprintf("%sSyncSuccessfull", strcase.ToCamel(objKindName))
+}
 
+// Sync mutates the subject of the syncer interface using controller-runtime
+// CreateOrUpdate method, when obj is not nil. It takes care of setting owner
+// references and recording kubernetes events where appropriate
+func Sync(ctx context.Context, syncer Interface, recorder record.EventRecorder) error {
+	result, err := syncer.Sync(ctx)
 	owner := syncer.GetOwner()
-	if recorder != nil && owner != nil {
-		reason := string(syncer.GetEventReasonForError(err))
-		if err != nil {
-			recorder.Eventf(owner, eventWarning, reason, "%T %s failed syncing: %s", obj, key, err)
-		}
-		if op != controllerutil.OperationResultNone {
-			recorder.Eventf(owner, eventNormal, reason, "%T %s %s successfully", obj, key, op)
+
+	if recorder != nil && owner != nil && result.EventType != "" && result.EventReason != "" && result.EventMessage != "" {
+		if result.Operation != controllerutil.OperationResultNone {
+			recorder.Eventf(owner, result.EventType, result.EventReason, result.EventMessage)
```

Comment:
In case `result.Operation == controllerutil.OperationResultNone`, but `err != nil`, nothing will be logged.